opm/autodiff/NewtonIterationBlackoilInterleaved.hpp

```diff
@@ -132,14 +132,37 @@ namespace Opm
                         AdditionalObjectDeleter<SeqPreconditioner> >
         constructPrecond(Operator& opA, const Comm& comm) const
         {
-            const double relax = 1.0;
-            SeqPreconditioner* seq_precond= new SeqPreconditioner(opA.getmat(),
-                                                                  relax);
             typedef AdditionalObjectDeleter<SeqPreconditioner> Deleter;
-            std::unique_ptr<ParPreconditioner, Deleter>
-                precond(new ParPreconditioner(*seq_precond, comm),
-                        Deleter(*seq_precond));
-            return precond;
+            typedef std::unique_ptr<ParPreconditioner, Deleter> Pointer;
+            int ilu_setup_successful = 1;
+            std::string message;
+            const double relax = 1.0;
+            SeqPreconditioner* seq_precond = nullptr;
+            try {
+                seq_precond = new SeqPreconditioner(opA.getmat(),
+                                                   relax);
+            }
+            catch ( Dune::MatrixBlockError error )
+            {
+                message = error.what();
+                std::cerr<<"Exception occured on process " <<
+                    comm.communicator().rank() << " during " <<
+                    "setup of ILU0 preconditioner with message: " <<
+                    message<<std::endl;
+                ilu_setup_successful = 0;
+            }
+            // Check whether there was a problem on some process
+            if ( comm.communicator().min(ilu_setup_successful) == 0 )
+            {
+                if ( seq_precond ) // not null if constructor succeeded
+                {
+                    // prevent memory leak
+                    delete seq_precond;
```

Comment:
I have tried to convert the code to using std::unique_ptr. For the
interleaved version with its fixed preconditioners I can make it work
but I have to use a unique_ptr in the AdditionalObjectDeleter, which
needs to know the type at compile time.

I run into problems with the CPRPreconditioner, though. As the type of
the preconditioner is chosen at runtime and ergo the
AdditionalObjectDeleter we are bound to using a shared_ptr as
unique_ptr needs to know the type of the deleter at compile time. Ergo
the shared_ptr needs to use a deleter with a unique_ptr inside which
does not compile for good reason.